Chapter 8/Lesson 4/homework/homework.js

```
@@ -0,0 +1,351 @@
+    }
+}
+
+function isString(value) {
+    let str = "must be a string."
+    try {
+        if (typeof value !== "string") {
+            throw new Error(str);
+        }
+    } catch (err) {
+        return str;
+    }
+}
+
+function hasLengthBetween(a, b) {
+    let str = "must be between 1 and 100 characters long."
+    return function (value) {
+        try {
+            if (value.length < a || value.length > b) {
+                throw new Error(str);
+            }
+        } catch (err) {
+            return str;
+        }
+    }
+}
+
+
+function isNumber(value) {
+    let str = "must be a number."
+    try {
+        if (typeof value !== "number") {
+            throw new Error(str);
+        }
+    } catch (err) {
+        return str;
+    }
+}
+
+function isGreaterThan(number) {
+    let str = "must be greater than 0."
+    return function (value) {
+        try {
+            if (value < number) {
+                throw new Error(str);
+            }
+        } catch (err) {
+            return str;
+        }
+    }
+}
+
+
+function composeValidators() {
+    let arr = [].concat(...arguments);
+    return function (fieldName) {
+        return function (value) {
+            let result;
+            for (let i = 0; i < arr.length; i++) {
+                if (result = arr[i](value)) {
+                    console.error(fieldName + " " + result);
+                    return result;
+                }
+            }
+        }
+    }
+}
+
+let validateName = composeValidators(
+    isRequired,
+    isString,
+    hasLengthBetween(2, 100)
+)("Name");
+
+
+let validateAge = composeValidators(
+    isNumber,
+    isGreaterThan(0)
+)("Age");
+
+
+function withValidation(arrWithFunctions) {
+    let result = true;
+    return function (obj) {
+        return function () {
+            let arrWithParameters = [].concat(...arguments);
+            if (arrWithFunctions.length != arrWithParameters.length) {
+                console.error("Error"); return;
```

Here is the place where you need to throw an error.